### PR TITLE
server: Fix NPE when migrating volume without template

### PR DESCRIPTION
**Pull request by gpordeus**
### Description

If you try to migrate a VM with volume and its template was removed, it results in a NullPointerException and the migration doesn't occur. 

If the template is already in both pools, the NPE happens in `getVolumeBackingFile`, when `_vmTemplateDao.findById()` returns null and it then tries to call from it. The fix was to check if it is null (and since it can only be null if it was removed, since `srcVolumeInfo` has `templateId`) and return null so it does a FullClone Migration.

If the template is not in both pools, the NPE happens in `copyTemplateToTargetFilesystemStorageIfNeeded`, when `TemplateObjectTO` tries to `getInstallPath()` on a null object. The fix was the same, check if `findById()` returns null and, if so, don't copy it.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?
- Uploaded two templates (A and B) to ACS, created one VM for each. Deleted B.
- Called migrateVirtualMachineWithVolume to migrate VM A on both host and storage pool.
- Called migrateVirtualMachineWithVolume to migrate VM B on both host and storage pool.
- Deleted A.
- Called migrateVirtualMachineWithVolume to migrate VM A, again.
All migrations were successful.

### CODE REVIEW

**Comment by weizhouapache** [comment]
is it applicable for 4.18 ?

**Comment by gpordeus** [comment]
> thanks @gpordeus for the fix (or enhancement even) . Do you think you can codify your testing by adding an integration test for these fixes to make sure we have no regression on those in the future?

@DaanHoogland I have added the integration tests.

> is it applicable for 4.18 ?

@weizhouapache Yes

File: test/integration/component/test_storage_motion_kvm.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+
+    def setUp(self):
+        self.apiclient = self.testClient.getApiClient()
+        self.dbclient = self.testClient.getDbConnection()
+        self.cleanup = []
+
+    def tearDown(self):
+        #Clean up, terminate the created ISOs
+        cleanup_resources(self.apiclient, self.cleanup)
+        return
```
Comment (DaanHoogland) [inline test/integration/component/test_storage_motion_kvm.py:153]:
```suggestion
    def tearDown(self):
        super(TestStorageMotionAndTemplate, self).tearDown()
```

File: test/integration/component/test_storage_motion_kvm.py [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+
+        self.debug("Creating vm with: %s" % template)
+        new_virtual_machine = VirtualMachine.create(
+            self.api_client,
+            self.services["small"],
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            templateid=template.id,
+            serviceofferingid=self.small_offering.id
+        )
```
Comment (DaanHoogland) [inline test/integration/component/test_storage_motion_kvm.py:330]:
```suggestion
        new_virtual_machine = VirtualMachine.create(
            self.api_client,
            self.services["small"],
            accountid=self.account.name,
            domainid=self.account.domainid,
            templateid=template.id,
            serviceofferingid=self.small_offering.id
        )
        self.cleanup.append(new_virtual_machine)
```

File: test/integration/component/test_storage_motion_kvm.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+        self.debug("Created vm %s" % new_virtual_machine)
+
+        self.debug("Deleting template: %s" % template)
+        self.cleanup.pop()
+        template.delete(self.apiclient)
```
Comment (DaanHoogland) [inline test/integration/component/test_storage_motion_kvm.py:335]:
```suggestion
        template.delete(self.apiclient)
        self.cleanup.remove(template)
```

File: test/integration/component/test_storage_motion_kvm.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+
+        cmd = migrateVirtualMachineWithVolume.migrateVirtualMachineWithVolumeCmd()
+        cmd.hostid = host.id
+        cmd.virtualmachineid = new_virtual_machine.id
+        cmd.migrateto = []
+        for volume, pool in migrateto.items():
+            cmd.migrateto.append({'volume':volume, 'pool':pool})
+        self.debug("hostid: %s, vmid: %s, migrateto: %s", cmd.hostid, cmd.virtualmachineid, cmd.migrateto)
+
+        self.apiclient.migrateVirtualMachineWithVolume(cmd)
```
Comment (DaanHoogland) [inline test/integration/component/test_storage_motion_kvm.py:402]: in base.py you can find a class `VirtualMachine`, which has a method `migrate_vm_with_volume`, it does basically the same.

**Review by DaanHoogland — COMMENTED** [review]
test code looks good, but uses some out of date constructs. suggestions added.

**Comment by DaanHoogland** [comment]
> > thanks @gpordeus for the fix (or enhancement even) . Do you think you can codify your testing by adding an integration test for these fixes to make sure we have no regression on those in the future?
> 
> @DaanHoogland I have added the integration test.

good work, i left some comments, but nothing serious.

> 
> > is it applicable for 4.18 ?
> 
> @weizhouapache Yes

can you rebase the change on the 4.18 branch than?

File: .asf.yaml [hunk 5]
```diff
@@ -49,11 +44,6 @@ github:
     rebase: false
 
   collaborators:
-    - kiranchavala
-    - rajujith
-    - alexandremattioli
-    - vishesh92
-    - soreana
```
Comment (DaanHoogland) [inline .asf.yaml]: Why do you want these contributors from this list?

**Comment by DaanHoogland** [comment]
@gpordeus all of a sudden all kinds of unrelated files are changed in your PR. Can you have a look?

**Comment by gpordeus** [comment]
> @gpordeus all of a sudden all kinds of unrelated files are changed in your PR. Can you have a look?

Fixed, but the labels added by the bot are still there.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
code looks good to me

**Comment by vladimirpetrov** [comment]
Hi @gpordeus, I tested and indeed there's no NPE anymore but the VMs are not able to start after a stop/start operation (on a KVM environment), did you notice the same behavior?
image

**Comment by gpordeus** [comment]
> Hi @gpordeus, I tested and indeed there's no NPE anymore but the VMs are not able to start after a stop/start operation (on a KVM environment), did you notice the same behavior? image

Hi, @vladimirpetrov ! I did not, could you specify in which circumstances this happens? If it's before or after deleting the template and/or migrating, for instance.

**Comment by vladimirpetrov** [comment]
This happens after deleting the template and successfully migrating the VM. Then when I do stop/start of the VM it doesn't start because of this error @gpordeus .

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]
LGTM but I haven't tested it

**Comment by DaanHoogland** [comment]
> > Hi @gpordeus, I tested and indeed there's no NPE anymore but the VMs are not able to start after a stop/start operation (on a KVM environment), did you notice the same behavior? image
> 
> Hi, @vladimirpetrov ! I did not, could you specify in which circumstances this happens? If it's before or after deleting the template and/or migrating, for instance.

@gpordeus (cc @vladimirpetrov ) any explanation/resolution on this?

**Comment by weizhouapache** [comment]
@gpordeus 
do you have any update ?
in my opinion, if template does not exist on secondary storage and primary storage, an exception should be thrown somewhere.

**Comment by gpordeus** [comment]
> @gpordeus (cc @vladimirpetrov ) any explanation/resolution on this?

@DaanHoogland , I've managed to reproduce the error (It happened when I migrated from an NFS primary storage to another), but I'm still looking into what's causing it. 


> @gpordeus do you have any update ? in my opinion, if template does not exist on secondary storage and primary storage, an exception should be thrown somewhere.

@weizhouapache I don't think an Exception should be thrown since the user is already allowed to force delete the template and the migration copies the template into the volume, so there is no absolute need for the backing file.

**Comment by weizhouapache** [comment]
> the migration copies the template into the volume

@gpordeus 
I hope so, can you confirm it ?

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by weizhouapache** [comment]
@gpordeus 
I can reproduce the issue @vladimirpetrov faced in his testing

- create a env with 2 primary pools, and disable pool primary2
- register a template
- create two vms (A and B)
- force-delete the template
- enable pool primary2
- migrate running vm A to primary2, it succeeds
- stop vm B, and migrate it to primary2, it succeeds
- start vm B, all looks good.
- stop vm A, and start it again, vm A is stuck at booting
image

the qcow2 image does not have a backing file
```
# qemu-img info -U /mnt/c2876c2e-5312-3cf9-b332-64d597086dc9/1a6593ad-8d02-4b10-a520-3eeefb336b13
image: /mnt/c2876c2e-5312-3cf9-b332-64d597086dc9/1a6593ad-8d02-4b10-a520-3eeefb336b13
file format: qcow2
virtual size: 50 MiB (52428800 bytes)
disk size: 1.06 MiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    compression type: zlib
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
    extended l2: false
```

**Comment by weizhouapache** [comment]
@DaanHoogland @rohityadavcloud @gpordeus 

my summary: if template is removed,

Before this PR,
- when migrate volume of stopped vm, it succeeds. vm also works after starting
- when migrate a running vm with volumes, got an NPE
Failed to migrate VM [VM instance {"id":3,"instanceName":"i-2-3-VM","type":"User","uuid":"1d3541d7-187e-417f-9cd4-cd18d0fc1e3d"}] along with its volumes due to [java.lang.NullPointerException].

With this PR:
- the NPE issue in live storage migration is fixed
- However, vm is corrupted after live storage migration

I would rather get an NPE exception than vm corruption , as data loss is a more critical issue for users.

**Review by weizhouapache — CHANGES_REQUESTED** [review]
@gpordeus 
can you please fix the vm corruption issue in our testing ?

**Comment by gpordeus** [comment]
> @gpordeus I can reproduce the issue @vladimirpetrov faced in his testing
> 
> image of vm B does not have backing file either, but it works (the disk size is much bigger than vm A)

@weizhouapache Great, thank you for testing and for pointing out the disk size. I hadn't been able to replicate it before [last week](https://github.com/apache/cloudstack/pull/7615#issuecomment-1684389067) because I was testing from and to iSCSI and NFS storages, while this error only happens in NFS to NFS.

> @DaanHoogland @rohityadavcloud @gpordeus
> 
> my summary: if template is removed,
> 
> Before this PR,
> 
>     * when migrate volume of stopped vm, it succeeds. vm also works after starting
> 
>     * when migrate a running vm with volumes, got an NPE
>       Failed to migrate VM [VM instance {"id":3,"instanceName":"i-2-3-VM","type":"User","uuid":"1d3541d7-187e-417f-9cd4-cd18d0fc1e3d"}] along with its volumes due to [java.lang.NullPointerException].
> 
> 
> With this PR:
> 
>     * the NPE issue in live storage migration is fixed
> 
>     * However, vm is corrupted after live storage migration
> 
> 
> I would rather get an NPE exception than vm corruption , as data loss is a more critical issue for users.

> @gpordeus can you please fix the vm corruption issue in our testing ?

Yes, I agree and I'm already working on it. The problem is a flag that is passed to libvirt migration only when it's NFS to NFS, for some reason, which makes it so that only the last layer of the vm is copied and, since the template was deleted and wasn't copied first, it corrupts them. Getting rid of the flag makes it work, but I'm testing if there are any other side effects or if there is a better way of going about it.

**Comment by DaanHoogland** [comment]
@gpordeus any progress? can we add this to 4.19?

**Comment by gpordeus** [comment]
> @gpordeus any progress? can we add this to 4.19?
@DaanHoogland Sorry for the delay, it's being a little trickier than I anticipated. It won't take much longer, but not enough to make it to 4.19.

**Comment by DaanHoogland** [comment]
@gpordeus when you get to this, will it include a fix for #5759 ?

**Comment by DaanHoogland** [comment]
@gpordeus should we still try to merge this? (as 4.18 is coming to end of maintemnance)

**Comment by gpordeus** [comment]
> @gpordeus should we still try to merge this? (as 4.18 is coming to end of maintemnance)

@DaanHoogland I suppose not, especially with #8911. I'm closing this, but I'll still be working on #5759. I was sidetracked, but I'll give it priority now.

**Pull request closed without merging**